src/lib/external_transition_validation/external_transition_validation.ml

```diff
@@ -0,0 +1,266 @@
+
+  let all t =
+    ( (`Time_received, t)
+    , (`Proof, t)
+    , (`Frontier_dependencies, t)
+    , (`Staged_ledger_diff, t) )
+
+  let fully_invalid = all Truth.False
+
+  let fully_valid = all Truth.True
+
+  module Unsafe = struct
+    let set_valid_time_received :
+           ( [`Time_received] * Truth.false_t
+           , 'proof
+           , 'frontier_dependencies
+           , 'staged_ledger_diff )
+           t
+        -> ( [`Time_received] * Truth.true_t
+           , 'proof
+           , 'frontier_dependencies
+           , 'staged_ledger_diff )
+           t = function
+      | ( (`Time_received, Truth.False)
+        , proof
+        , frontier_dependencies
+        , staged_ledger_diff ) ->
+          ( (`Time_received, Truth.True)
+          , proof
+          , frontier_dependencies
+          , staged_ledger_diff )
+      | _ -> failwith "why can't this be refuted?"
+
+    let set_valid_proof :
+           ( 'time_received
+           , [`Proof] * Truth.false_t
+           , 'frontier_dependencies
+           , 'staged_ledger_diff )
+           t
+        -> ( 'time_received
+           , [`Proof] * Truth.true_t
+           , 'frontier_dependencies
+           , 'staged_ledger_diff )
+           t = function
+      | ( time_received
+        , (`Proof, Truth.False)
+        , frontier_dependencies
+        , staged_ledger_diff ) ->
+          ( time_received
+          , (`Proof, Truth.True)
+          , frontier_dependencies
+          , staged_ledger_diff )
+      | _ -> failwith "why can't this be refuted?"
+
+    let set_valid_frontier_dependencies :
+           ( 'time_received
+           , 'proof
+           , [`Frontier_dependencies] * Truth.false_t
+           , 'staged_ledger_diff )
+           t
+        -> ( 'time_received
+           , 'proof
+           , [`Frontier_dependencies] * Truth.true_t
+           , 'staged_ledger_diff )
+           t = function
+      | ( time_received
+        , proof
+        , (`Frontier_dependencies, Truth.False)
+        , staged_ledger_diff ) ->
+          ( time_received
+          , proof
+          , (`Frontier_dependencies, Truth.True)
+          , staged_ledger_diff )
+      | _ -> failwith "why can't this be refuted?"
+
+    let set_valid_staged_ledger_diff :
+           ( 'time_received
+           , 'proof
+           , 'frontier_dependencies
+           , [`Staged_ledger_diff] * Truth.false_t )
+           t
+        -> ( 'time_received
+           , 'proof
+           , 'frontier_dependencies
+           , [`Staged_ledger_diff] * Truth.true_t )
+           t = function
+      | ( time_received
+        , proof
+        , frontier_dependencies
+        , (`Staged_ledger_diff, Truth.False) ) ->
+          ( time_received
+          , proof
+          , frontier_dependencies
+          , (`Staged_ledger_diff, Truth.True) )
+      | _ -> failwith "why can't this be refuted?"
+  end
+
+  let validate_time_received (t, validation) ~time_received =
```

Comment:
don't you think we should bind the validation tighter to the `External_transition`?

Like `External_transition.t` should be:

```ocaml
type t_ = (* whatever external_transition is now *)
type ('p1, 'p2, ..., 'pn) t =
  (('p1, 'p2) validation * t_)
```

and all functions on External_transition act on `t` (sometimes requiring validation as needed).

Basically make it impossible for people to consume an External_transition without thinking about the validation. If it's just a tuple, someone may use External_transition without it.

Reply:
I think it's better to provide a separate module so that we can fully encapsulate. If all this logic is moved into `External_transition`, then `Unsafe` functions would need to be surfaced since the `External_transition` cannot define the functions needed for interacting with the transition frontier and the staged ledger. I don't think there are any functions in `External_transition` that would benefit from talking about the validation either. The idea is that all the high level components that consume and pass external transitions around will talk about them through the `External_transition_validation.with_transition` type before finally lowering the external transition down into a `Breadcrumb.t` once the type unifies with `External_transition_validation.fully_valid`.

Reply:
Okay this makes sense, but we should still keep the `External_transition.Verified.t` as discussed.